### DISCUSSION TITLE
vscode rust devcontainer: cargo build resulting in "failed to link or copy

### DISCUSSION THREAD

**Discussion by dss010101** [devcontainers/community#141] [General]
### Problem

Im not sure if should be reaching out here or vscode for help on this...but i will start here and let  me know if im int he wrong location..

i have built a vs code devcontainer based on ```mcr.microsoft.com/devcontainers/rust:1-1-bullseye```, running under Docker Desktop on windows 11.  I am able to run a basic example using this without any dependencies

i have the following structure (Note: the target and Cargo.lock do not exist until after i attempt a ```cargo build```:

[![enter image description here][1]][1]

  [1]: https://i.sstatic.net/ts9wsRyf.png

My toml looks like this:

```
[package]
name = "aipg_client"
version = "0.1.0"
edition = "2021"

#[dependencies]
#tokio-postgres = "0.7.10"
```

when i do ```Cargo build``` or run from command line, i see this:

```
vscode ➜ /workspaces/aipg_client $ cargo clean
     Removed 19 files, 3.7MiB total
vscode ➜ /workspaces/aipg_client $ cargo build
   Compiling aipg_client v0.1.0 (/workspaces/aipg_client)
warning: error copying object file `/workspaces/aipg_client/target/debug/deps/aipg_client-29a466bd116893b1.2wavitxjlghb5boo.rcgu.o` to incremental directory as `/workspaces/aipg_client/target/debug/incremental/aipg_client-cn2ostj37iq8/s-gw2y4x6qjv-d1xpw0-working/2wavitxjlghb5boo.o`: Operation not permitted (os error 1)

warning: error copying object file `/workspaces/aipg_client/target/debug/deps/aipg_client-29a466bd116893b1.340xt2kclbcozwj5.rcgu.o` to incremental directory as `/workspaces/aipg_client/target/debug/incremental/aipg_client-cn2ostj37iq8/s-gw2y4x6qjv-d1xpw0-working/340xt2kclbcozwj5.o`: Operation not permitted (os error 1)

warning: error copying object file `/workspaces/aipg_client/target/debug/deps/aipg_client-29a466bd116893b1.3nlg3iffbximniu8.rcgu.o` to incremental directory as `/workspaces/aipg_client/target/debug/incremental/aipg_client-cn2ostj37iq8/s-gw2y4x6qjv-d1xpw0-working/3nlg3iffbximniu8.o`: Operation not permitted (os error 1)

warning: error copying object file `/workspaces/aipg_client/target/debug/deps/aipg_client-29a466bd116893b1.57xak5150to06626.rcgu.o` to incremental directory as `/workspaces/aipg_client/target/debug/incremental/aipg_client-cn2ostj37iq8/s-gw2y4x6qjv-d1xpw0-working/57xak5150to06626.o`: Operation not permitted (os error 1)

warning: error copying object file `/workspaces/aipg_client/target/debug/deps/aipg_client-29a466bd116893b1.ele9p2lo4vgyfma.rcgu.o` to incremental directory as `/workspaces/aipg_client/target/debug/incremental/aipg_client-cn2ostj37iq8/s-gw2y4x6qjv-d1xpw0-working/ele9p2lo4vgyfma.o`: Operation not permitted (os error 1)

warning: error copying object file `/workspaces/aipg_client/target/debug/deps/aipg_client-29a466bd116893b1.eyyxtxgek58rknf.rcgu.o` to incremental directory as `/workspaces/aipg_client/target/debug/incremental/aipg_client-cn2ostj37iq8/s-gw2y4x6qjv-d1xpw0-working/eyyxtxgek58rknf.o`: Operation not permitted (os error 1)

warning: `aipg_client` (bin "aipg_client") generated 6 warnings
error: failed to link or copy `/workspaces/aipg_client/target/debug/deps/aipg_client-29a466bd116893b1` to `/workspaces/aipg_client/target/debug/aipg_client`

Caused by:
  Operation not permitted (os error 1)
```

i checked these folders and they are all r/w.
What am i doing wrong?

### Steps

1. Create a vscode dev container:  CTRL+SHIFT+P -> devcontainer:new -> rust
2. Note: i jused a Dockerfile + Docker Compose along with .devcontainer.json. but the issue is reproducible with just a .devcontainer.json file
3. Install rust analyzer extension in the container 
4. from bash terminal: ```cargo new test```
5. run ```cargo build``` twice

### Version
rustc 1.78.0 (9b00956e5 2024-04-29)
cargo 1.78.0 (54d8815d0 2024-03-26)

vscode ➜ /workspaces/aipg_client $ cargo -vV
cargo 1.78.0 (54d8815d0 2024-03-26)
release: 1.78.0
commit-hash: 54d8815d04fa3816edc207bbc4dd36bf18014dbc
commit-date: 2024-03-26
host: x86_64-unknown-linux-gnu
libgit2: 1.7.2 (sys:0.18.2 vendored)
libcurl: 8.6.0-DEV (sys:0.4.72+curl-8.6.0 vendored ssl:OpenSSL/1.1.1w)
ssl: OpenSSL 1.1.1w  11 Sep 2023
os: Debian 11 (bullseye) [64-bit]

vscode ➜ /workspaces/aipg_client $ cat /etc/debian_version 
11.9
docker desktop v4.29.0

**Comment by dss010101** [comment]
though i dont know if its relevant here i are my docker file and docker-compose:
docker file:
```
FROM mcr.microsoft.com/devcontainers/rust:1-1-bullseye as aipg_client_base
ENV TZ=America/New_York

ARG ENV
ARG USER_ID
ARG GROUP_ID
ARG USER_NAME
ARG APP_HOME="/apps/aipg_client_base"
RUN echo "env: $ENV, user: $USER_NAME, user_id: $USER_ID, app_home: $APP_HOME"

RUN apt-get update && apt-get -y install acl iputils-ping less libpq-dev nano postgresql-client sudo

RUN groupadd -g $GROUP_ID -o $USER_NAME
RUN useradd -m -s /bin/bash -u $USER_ID -g $GROUP_ID $USER_NAME

FROM aipg_client_base as aipg_client_local


#attempt at handling this target folder issue...
RUN mkdir -p /workspaces/aipg_client/target
RUN chown -R vscode:vscode /workspaces/aipg_client/target
RUN chmod -R 777 /workspaces/aipg_client/target
```

docker-compose:
```
networks:
  aipg_client_net:
    driver: bridge

services:
  aipg_client:
    container_name: aipg_client_${ENV}
    image: "aipg_client:${VERSION}"
    build:
      context: ..
      dockerfile: .devcontainer/Dockerfile
      target: aipg_client_local
      args:
        ENV: ${ENV:-local}
        USER_ID: ${USER_ID:-1000}
        GROUP_ID: ${GROUP_ID:-1000}
        USER_NAME: ${USER_NAME}
      shm_size: '2gb'
    env_file:
      - ./.env
    volumes:
      - ../..:/workspaces
      - ~/.ssh:/home/${USER_NAME}/.ssh
      - //var/run/docker.sock:/var/run/docker.sock
    command: sleep infinity
    shm_size: 0.94gb
    ports: 
      - "${SVC_PORT}:5000"    
    networks:
      - aipg_client_net
```

.env file:
```
VERSION=local
ENV=local
USER_ID=197609
#USER_ID=1001
GROUP_ID=1000
USER_NAME=user1
SVC_PORT=6000
```

if i add a .cargo/config.toml with the target anywhere outside my project workspace, it seems to work fine
```
[build]
target-dir = "/tmp/target"
```

But ofcourse that means anytime i restart/rebuild the container, i have to rebuild the project/binaries.  which makes it not seem like a good solution

**Comment by dss010101** [comment]
btw - is there an official place for vscode/devcontainer support/issues?

**Reply by samruddhikhandale** [reply]
Opening a discussion is acceptable for support request. However, based on the type of problem (CLI, spec, images, Templates etc) you can open issues in the designated repositories. 

For the `Dev containers` extension support, you can open issues in https://github.com/microsoft/vscode-remote-release/issues

**Comment by samruddhikhandale** [comment]
Hi 👋 

The error message "Operation not permitted (os error 1)" suggests that the Rust compiler is unable to write to the target directory. This is likely due to a file system permission issue.

Given that you're running your Rust project inside a Docker container on Windows, one possible cause of this issue is the file sharing settings of Docker Desktop. Docker Desktop needs permission to share files between your Windows file system and the Docker containers.

Here are the steps to check and update your file sharing settings:

1. Open Docker Desktop.
2. Go to Settings > Resources > File Sharing.
3. Make sure the drive where your Rust project is located is checked.
4. If it's not, check it and click Apply & Restart

Let us know if this helps, else we can try digging more into the issue, thanks!

**Comment by dss010101** [comment]
> Hi 👋
> 
> The error message "Operation not permitted (os error 1)" suggests that the Rust compiler is unable to write to the target directory. This is likely due to a file system permission issue.
> 
> Given that you're running your Rust project inside a Docker container on Windows, one possible cause of this issue is the file sharing settings of Docker Desktop. Docker Desktop needs permission to share files between your Windows file system and the Docker containers.
> 
> Here are the steps to check and update your file sharing settings:
> 
> 1. Open Docker Desktop.
> 2. Go to Settings > Resources > File Sharing.
> 3. Make sure the drive where your Rust project is located is checked.
> 4. If it's not, check it and click Apply & Restart
> 
> Let us know if this helps, else we can try digging more into the issue, thanks!

Thank you.  i feel this may be on the track to what the issue is related to (especially considering my files are on a locally mapped google drive), but i do not find 'File Sharing' under the settings you mentioned.   I am using WSL2, and do see from the docs here that this option is not available when using WSL2: https://docs.docker.com/desktop/synchronized-file-sharing/
